### PR TITLE
UNDERTOW-1476: Fix URLDecodingHandler when query string is empty

### PR DESCRIPTION
**Pull request by cakofony**
Added test cases to prevent future regressions.

This PR is the most important subset of #710

### CODE REVIEW

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2642](https://ci.wildfly.org/viewLog.html?buildId=135664&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of bb550b98c76fb029cb1bd4cddb0f8af8a607892b
Summary: Tests failed: 1 (1 new), passed: 3362, ignored: 709 Build time: 00:43:26
### Failed tests
```
io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed[proxy][http2]: java.lang.AssertionError
	at io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed(ReceiverTestCase.java:194)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)
------- Stderr: -------
05:08:50,434 DEBUG (XNIO-1 I/O-1) [io.undertow.server.handlers.proxy] <ProxyHandler.java:549> Sending request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target localhost/127.0.0.1:8888 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
05:08:50,434 DEBUG (XNIO-1 I/O-1) [io.undertow.server.handlers.proxy] <ProxyHandler.java:566> Sent request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target 127.0.0.1 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
05:08:50,489 ERROR (XNIO-1 I/O-1) [io.undertow.proxy] <ProxyHandler.java:662> UT005028: Proxy request to /fullbytes failed: java.nio.channels.ClosedChannelException
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:99)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:1061)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:975)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.terminated(SslConduit.java:1214)
	at org.xnio.nio.NioSocketConduit.readTerminated(NioSocketConduit.java:360)
	at org.xnio.nio.NioSocketStreamConnection.notifyReadClosed(NioSocketStreamConnection.java:148)
	at org.xnio.Connection.close(Connection.java:139)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.ssl.SslConduit.closed(SslConduit.java:1046)
	at io.undertow.protocols.ssl.SslConduit.close(SslConduit.java:1117)
	at io.undertow.protocols.ssl.UndertowSslConnection.closeAction(UndertowSslConnection.java:149)
	at org.xnio.Connection.close(Connection.java:132)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.close(AbstractFramedChannel.java:816)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:832)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:809)
	at io.undertow.client.http2.Http2ClientConnection.close(Http2ClientConnection.java:303)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.handlers.proxy.ProxyHandler$IoExceptionHandler.handleException(ProxyHandler.java:790)
	at org.xnio.ChannelListeners.invokeChannelExceptionHandler(ChannelListeners.java:126)
	at io.undertow.util.Transfer$TransferListener.handleEvent(Transfer.java:218)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel$1.run(AbstractFramedStreamSinkChannel.java:214)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:935)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:929)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at org.xnio.conduits.ReadReadyHandler$ChannelListenerHandler.readReady(ReadReadyHandler.java:66)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1175)
	at io.undertow.protocols.ssl.SslConduit$1.run(SslConduit.java:168)
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466)

05:08:50,489 DEBUG (XNIO-1 I/O-1) [io.undertow.request.error-response] <DelegatingBasicLogger.java:424> Setting error code 503 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}: java.lang.RuntimeException
	at io.undertow.server.HttpServerExchange.setStatusCode(HttpServerExchange.java:1410)
	at io.undertow.server.handlers.proxy.ProxyHandler.handleFailure(ProxyHandler.java:668)
	at io.undertow.server.handlers.proxy.ProxyHandler$ResponseCallback.failed(ProxyHandler.java:727)
	at io.undertow.client.http2.Http2ClientExchange.failed(Http2ClientExchange.java:125)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)


```

**Comment by undertow-pull-request** [comment]
**Linux** [Build 2892](https://ci.wildfly.org/viewLog.html?buildId=135762&buildTypeId=UT_MASTER_LINUX) outcome was **UNKNOWN** using a merge of acfc4ecfe13e45e77b76ce040124832e090cdea7
Summary: Canceled (Error while applying patch; cannot find commit dfde27f66f3dcee668e5c72fe08c13161afb5a3e in the git://github.com/undertow-io/undertow.git repository, possible reason: refs/pull/712/merge branch was updated and the commit selected for the bui... Build time: 00:00:09

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2643](https://ci.wildfly.org/viewLog.html?buildId=135763&buildTypeId=UNDERTOW_MasterWindows) outcome was **UNKNOWN** using a merge of acfc4ecfe13e45e77b76ce040124832e090cdea7
Summary: Canceled (Error while applying patch; cannot find commit dfde27f66f3dcee668e5c72fe08c13161afb5a3e in the git://github.com/undertow-io/undertow.git repository, possible reason: refs/pull/712/merge branch was updated and the commit selected for the bui... Build time: 00:00:13

File: core/src/main/java/io/undertow/server/handlers/URLDecodingHandler.java [hunk 1]
```diff
@@ -73,13 +73,13 @@ public void handleRequest(final HttpServerExchange exchange) throws Exception {
...
-                        for (Map.Entry<String, String> entry : parameters.entrySet()) {
-                            entry.setValue(URLUtils.decode(entry.getValue(), charset, true, true, sb));
-                        }
+            }
+            PathTemplateMatch pathTemplateMatch = exchange.getAttachment(PathTemplateMatch.ATTACHMENT_KEY);
+            if (pathTemplateMatch != null) {
+                Map<String, String> parameters = pathTemplateMatch.getParameters();
+                if (parameters != null) {
+                    for (Map.Entry<String, String> entry : parameters.entrySet()) {
+                        entry.setValue(URLUtils.decode(entry.getValue(), charset, true, false, sb));
```
Comment (cakofony) [inline core/src/main/java/io/undertow/server/handlers/URLDecodingHandler.java]: Updated formEncoding from true to false + added a test

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2644](https://ci.wildfly.org/viewLog.html?buildId=135766&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 980a4c270c2c3083c711aaa3f281892fba922873
Summary: Tests failed: 32 (27 new), passed: 3335, ignored: 697 Build time: 00:42:33
### Failed tests
```
io.undertow.server.HttpServerExchangeTestCase.testHttpServerExchange[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.HttpServerExchangeTestCase.testHttpServerExchange(HttpServerExchangeTestCase.java:66)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.NewlineInHeadersTestCase.testNewlineInHeaders[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.NewlineInHeadersTestCase.testNewlineInHeaders(NewlineInHeadersTestCase.java:63)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.AllowedMethodsTestCase.testDisallowedMethods[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.AllowedMethodsTestCase.testDisallowedMethods(AllowedMethodsTestCase.java:74)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.ChunkedRequestNotConsumedTestCase.testChunkedRequestNotConsumed[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.ChunkedRequestNotConsumedTestCase.testChunkedRequestNotConsumed(ChunkedRequestNotConsumedTestCase.java:121)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)
------- Stdout: -------
Using Seed -259239655


io.undertow.server.handlers.HttpContinueConduitWrappingHandlerTestCase.testHttpContinueAcceptedWithChunkedRequest[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.HttpContinueConduitWrappingHandlerTestCase.testHttpContinueAcceptedWithChunkedRequest(HttpContinueConduitWrappingHandlerTestCase.java:151)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.ExceptionHandlerTestCase.testExceptionMappers[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.ExceptionHandlerTestCase.testExceptionMappers(ExceptionHandlerTestCase.java:80)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.JDBCLogDatabaseTestCase.testSingleLogMessageToDatabase[proxy][http2-clear-upgrade]: org.h2.jdbc.JdbcSQLException: No data is available [2000-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.message.DbException.get(DbException.java:144)
	at org.h2.jdbc.JdbcResultSet.checkOnValidRow(JdbcResultSet.java:3218)
	at org.h2.jdbc.JdbcResultSet.get(JdbcResultSet.java:3231)
	at org.h2.jdbc.JdbcResultSet.get(JdbcResultSet.java:3248)
	at org.h2.jdbc.JdbcResultSet.getString(JdbcResultSet.java:310)
	at io.undertow.server.handlers.JDBCLogDatabaseTestCase.testSingleLogMessageToDatabase(JDBCLogDatabaseTestCase.java:153)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.AllowedMethodsTestCase.testAllowedMethods[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.AllowedMethodsTestCase.testAllowedMethods(AllowedMethodsTestCase.java:52)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.HttpContinueAcceptingHandlerTestCase.testHttpContinueRejected[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.HttpContinueAcceptingHandlerTestCase.testHttpContinueRejected(HttpContinueAcceptingHandlerTestCase.java:103)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.ChunkedResponseTransferCodingTestCase.sendHttpRequest[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.ChunkedResponseTransferCodingTestCase.sendHttpRequest(ChunkedResponseTransferCodingTestCase.java:88)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)


io.undertow.server.handlers.FixedLengthRequestTestCase.testFixedLengthRequest[proxy][http2-clear-upgrade]: java.net.SocketException: No buffer space available (maximum connections reached?): connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:326)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at io.undertow.server.handlers.FixedLengthRequestTestCase.testFixedLengthRequest(FixedLengthRequestTestCase.java:97)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)



##### there are 22 more failed tests, see build details
```

**Comment by cakofony** [comment]
Closing in favor of #710

**Pull request closed without merging**